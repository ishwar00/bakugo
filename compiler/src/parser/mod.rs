use miette::NamedSource;
use pest::{
    error::Error,
    iterators::{Pair, Pairs},
    Parser,
};

use crate::ast::{BakugoParsingErrorDisplay, Node, SourceFile};

#[derive(pest_derive::Parser)]
#[grammar = "bakugo.pest"]
pub struct BakugoParser;

pub fn parse_string(s: &str) -> Result<Pairs<'_, Rule>, Box<Error<Rule>>> {
    match BakugoParser::parse(Rule::SourceFile, s) {
        Ok(ast) => Ok(ast),
        Err(err) => Err(Box::new(err)),
    }
}

pub fn construct_ast(
    source: NamedSource,
    pair: Pair<'_, Rule>,
) -> Result<SourceFile<'_>, BakugoParsingErrorDisplay> {
    match SourceFile::parse(pair) {
        Ok(s) => Ok(s),
        Err(bakugo_err) => Err(BakugoParsingErrorDisplay::from_error(bakugo_err, source)),
    }
}
