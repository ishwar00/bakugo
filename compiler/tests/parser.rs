use std::fs;

use bakugo::parser::parse_string;
use insta::{assert_snapshot, assert_yaml_snapshot, glob};

#[test]
fn test_parser() {
    glob!("examples/*.bakugo", |path| {
        let input = fs::read_to_string(path).unwrap();
        let parsed = parse_string(&input);
        match parsed {
            Ok(parsed) => assert_yaml_snapshot!(parsed),
            Err(err) => assert_snapshot!(err.to_string()),
        }
    });
}
