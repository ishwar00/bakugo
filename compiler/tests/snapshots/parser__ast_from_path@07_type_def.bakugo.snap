---
source: tests/parser.rs
expression: parsed
input_file: tests/examples/07_type_def.bakugo
---
SourceFile {
    top_level: [
        FnDecl(
            FnDecl {
                name: Ident {
                    value: "main",
                    span: Span {
                        str: "main",
                        start: 5,
                        end: 9,
                    },
                },
                params: [],
                result_kind: None,
                body: StatementList(
                    [
                        Declaration(
                            Kind(
                                KindDecl {
                                    name: Ident {
                                        value: "a",
                                        span: Span {
                                            str: "a",
                                            start: 20,
                                            end: 21,
                                        },
                                    },
                                    kind: Simple {
                                        name: "int",
                                        span: Span {
                                            str: "int",
                                            start: 22,
                                            end: 25,
                                        },
                                    },
                                },
                            ),
                        ),
                        Declaration(
                            Kind(
                                KindDecl {
                                    name: Ident {
                                        value: "b",
                                        span: Span {
                                            str: "b",
                                            start: 37,
                                            end: 38,
                                        },
                                    },
                                    kind: Simple {
                                        name: "string",
                                        span: Span {
                                            str: "string",
                                            start: 39,
                                            end: 45,
                                        },
                                    },
                                },
                            ),
                        ),
                        Declaration(
                            Kind(
                                KindDecl {
                                    name: Ident {
                                        value: "c",
                                        span: Span {
                                            str: "c",
                                            start: 49,
                                            end: 50,
                                        },
                                    },
                                    kind: Simple {
                                        name: "float",
                                        span: Span {
                                            str: "float",
                                            start: 51,
                                            end: 56,
                                        },
                                    },
                                },
                            ),
                        ),
                    ],
                ),
                span: Span {
                    str: "func main() {\n\ttype a int;\n\ttype (\n\t\tb string;\n\t\tc float;\n\t);\n}",
                    start: 0,
                    end: 63,
                },
            },
        ),
    ],
}
